skfda/misc/metrics/_sdtw_distances.py

```diff
@@ -0,0 +1,368 @@
+        )
+
+    if not (fdata1.n_samples == 1 and fdata2.n_samples == 1):
+        raise ValueError(
+            "Both fdata1 and fdata2 must each contain a single sample"
+        )
+
+
+def half_sq_euclidean(
+    arg1: NDArrayFloat,
+    arg2: Optional[NDArrayFloat] = None,
+) -> NDArrayFloat:
+    """
+    Return the half squared Euclidean row distance between
+    two matrices of size (n1, d) and (n2, d).
+
+    Args:
+        arg1: First sample (n_samples_1, n_features).
+        arg2: Second sample (n_samples_2, n_features).
+            If it is ``None`` returns the half squared Euclidean distance
+            between the rows of ``arg1``.
+
+    The distance is taken w.r.t matrix rows.
+    """
+    if arg2 is not None:
+        # sq_euclidean(X, Y) =
+        # DotProduct(sigma_0_bounds=(10**-10, 10**10)) =
+        # diag(np.dot(X, X.T)) + diag(np.dot(Y, Y.T)) - 2*np.dot(X, Y.T)
+        # np.dot(X, Y.T)=inner_product_matrix(X, Y)
+        cost_12 = -1 * inner_product_matrix(arg1, arg2)
+
+        # np.sum(X ** 2, axis=1)=inner_product(X, X)
+        cost_12 += (0.5 * inner_product(arg1, arg1))[:, np.newaxis]
+        cost_12 += 0.5 * inner_product(arg2, arg2)
+
+        return cost_12
+    else:
+        cost_11 = -1 * inner_product_matrix(arg1, arg2)
+        # half Sum Of Squares on each row
+        sos_row = 0.5 * inner_product(arg1, arg1)
+        cost_11 += sos_row[:, np.newaxis]
+        cost_11 += sos_row
+
+        return cost_11
+
+
+def _sdtw_divergence(
+    fdata1: FDataGrid,
+    fdata2: FDataGrid,
+    *,
+    gamma: float = 1.0,
+    cost: Union[Tuple[NDArrayFloat], Callable[[
+        NDArrayFloat, NDArrayFloat, NDArrayFloat]]] = None,
+    check_cost: bool = False,
+    **kwargs_cost: Dict
+) -> float:
+
+    # fdata1 and fdata2 are only required to be one-dimensional.
+    # They can have different (one-dimensional) domain ranges and
+    # different grid points size
+
+    _check_input_fdata(fdata1, fdata2)
+
+    n1, n2 = len(fdata1.grid_points[0]), len(fdata2.grid_points[0])
+
+    if gamma <= 0:
+        raise ValueError(
+            f"gamma was set to {gamma} but must be positive"
+        )
+
+    # booleans to remember the input format of cost if given
+    cost_was_tuple = isinstance(cost, tuple)
+    cost_was_callable = isinstance(cost, Callable)
+
+    if cost_was_tuple:
+        if len(cost) != 3:
+            raise ValueError(
+                "When the alignment cost matrices are pre-computed, "
+                "then 'cost' must be a tuple of three two-dimensional "
+                "numpy arrays."
+            )
+
+    elif cost_was_callable:
+        # pre-computation for the symmetry check
+        # of the cross-product part of the cost
+        if check_cost:
+            cost_21 = cost(
+                fdata2.data_matrix[0],
+                fdata1.data_matrix[0],
+                **kwargs_cost
+            )
+
+        # compute the cost matrices
+        cost = (
+            cost(
```

Comment:
- I think that the cost function should receive the `FDataGrid`s instead of the matrices. This method arose in time-series, where grid points are usually equispaced and each series could be considered a vector. That is not necessarily the case in FDA. It is possible for a `FDataGrid` to have non-equispaced grid points, and this is relevant, for example, when computing the inner products and the distances between functions, as each point will have a different weight  in the quadrature.
- Also, are you using only one sample of each FDataGrid? In scikit-fda distances are paired (it computes the distance between the i-th element of the first dataset and the i-th element of the second one) except if one dataset has only one sample, in which case it is "broadcasted". Usually you can achieve this just by keeping all the observations and using vectorized NumPy operations.